TF-7749 Update Policy Sets API documentation (#423)
hashicorp/terraform-docs-common: website/docs/cloud-docs/api-docs/policy-sets.mdx

---
page_title: Policy Sets - API Docs - Terraform Cloud
description: >-

>   Use the `/policy-sets` endpoint to manage groupings of Sentinel and OPA policies. List, show, create, and update policy sets and policy set versions, and attach and detach policies from workspaces, projects, and exclusions using the HTTP API.

Reviewer: I wonder if there is a better way to word this whole thing 🤔 I'll think about this more! It's getting a bit unruly and finding it hard to understand. Rose might have some thoughts!

Author: @rkoron007

[Policy Enforcement](/terraform/cloud-docs/policy-enforcement) lets you use the policy-as-code frameworks Sentinel and Open Policy Agent (OPA) to apply policy checks to Terraform Cloud workspaces.

> [Policy sets](/terraform/cloud-docs/policy-enforcement/manage-policy-sets) are collections of policies that you can apply globally or to specific [projects](/terraform/cloud-docs/workspaces/organize-workspaces-with-projects) and workspaces, except the ones in the exclusions list. For each run in the selected workspaces, Terraform Cloud checks the Terraform plan against the policy set.

Reviewer 2: I agree with Keran, I don't think we should mention exclusions here because we are still explaining what a policy set is. We want folks to have that firm understanding before building on that knowledge:
```suggestion
[Policy sets](/terraform/cloud-docs/policy-enforcement/manage-policy-sets) are collections of policies that you can apply globally or to specific [projects](/terraform/cloud-docs/workspaces/organize-workspaces-with-projects) and workspaces. For each run in the selected workspaces, Terraform Cloud checks the Terraform plan against the policy set.
```

| `data.relationships.projects.data[]`          | array\[object] | `[]`      | A list of resource identifier objects that defines which projects are associated with the policy set. These objects must contain `id` and `type` properties, and the `type` property must be `projects`. For example, `{ "id": "prj-AwfuCJTkdai4xj9w", "type": "projects" }`.
| `data.relationships.workspaces.data[]`        | array\[object] | `[]`      | A list of resource identifier objects that defines which workspaces are associated with the policy set. These objects must contain `id` and `type` properties, and the `type` property must be `workspaces`. For example, `{ "id": "ws-2HRvNs49EWPjDqT1", "type": "workspaces" }`. Obtain workspace IDs from the [workspace's settings page](/terraform/cloud-docs/workspaces/settings) or the [Show Workspace endpoint](/terraform/cloud-docs/api-docs/workspaces#show-workspace). You can only specify this attribute when `data.attributes.global` is `false`. |

> | `data.relationships.exclusions.data[]`        | array\[object] | `[]`      | A list of resource identifier objects that defines which excluded workspaces and are associated with the policy set. These objects must contain `id` and `type` properties, and the `type` property must be `workspaces`. For example, `{ "id": "ws-FVVvzCDaykN1oHiw", "type": "workspaces" }`.                                                                                                                                                                                                                                               |

Reviewer 2: Adding to Keran's above suggestion with a teeny tweak for active voice:
```suggestion
| `data.relationships.exclusions.data[]`        | array\[object] | `[]`      | A list of resource identifier objects specifying which workspaces Terraform Cloud excludes from a policy set's enforcement. These objects must contain `id` and `type` properties, and the `type` property must be `workspaces`. For example, `{ "id": "ws-FVVvzCDaykN1oHiw", "type": "workspaces" }`.                                                                                                                                                                                                                                               |
```

}
```

> -> **Note:** The `data.relationships.projects`, `data.relationships.workspaces`, and `data.relationships.exclusions` objects refer to projects, workspaces and exclusions directly attached to the policy set. These keys are omitted for policy sets marked as global, which are implicitly related to all of the organization's workspaces.

Reviewer 2: Can we split this out into two sentences? I think `exclusions` are different enough that they need more explanation:
```suggestion
-> **Note:** The `data.relationships.projects` and `data.relationships.workspaces` refer to the projects and workspaces attached to the policy set. The `data.relationships.exclusions` object refers to the attached workspaces that Terraform Cloud excludes from a policy set's rules. Terraform Clouds omits these keys for policy sets marked as global, which are implicitly related to all of the organization's workspaces.
```

| `data.relationships.projects`                 | array\[object] | (previous value) | An array of references to projects that the policy set should be assigned to. Sending an empty array clears all project assignments. You can only specify this attribute when `data.attributes.global` is `false`.                                                                                           |                                                                                                                                          |
| `data.relationships.workspaces`               | array\[object] | (previous value) | An array of references to workspaces that the policy set should be assigned to. Sending an empty array clears all workspace assignments. You can only specify this attribute when `data.attributes.global` is `false`.                                                                                       | 

> | `data.relationships.exclusions`               | array\[object] | (previous value) | An array of references to workspaces that the policy set should not be assigned to. Sending an empty array clears all exclusions assignments. You can only specify this attribute when `data.attributes.global` is `false`.                                                                                      |                                                                                                                                          |

Reviewer 2: Question: is it that Terraform Cloud doesn't _assign_ the policy set? Or that TFC doesn't _run_ the policy set for excluded workspaces? 

Small suggestion for active voice, but again not sure if this is correct without the above question:
```suggestion
| `data.relationships.exclusions`               | array\[object] | (previous value) | An array of references to workspaces that Terraform Cloud excludes from a policy set. Sending an empty array clears all exclusions assignments. You can only specify this attribute when `data.attributes.global` is `false`.                                                                                      |                                                                                                                                          |
```

Author: Its this one: Or that TFC doesn't run the policy set for excluded workspaces

Reviewer 2: Thank you for clarifying! With that background,  I think this feels a little more clear (feel free to push back if you don't think so though!):
```suggestion
| `data.relationships.exclusions`               | array\[object] | (previous value) | An array of references to excluded workspaces that Terraform Cloud will not enforce this policy set upon. Sending an empty array clears all exclusions assignments. You can only specify this attribute when `data.attributes.global` is `false`.                                                                                      |                                                                                                                                          |
```

Author: I have changed it to: An array of references to excluded workspaces that the policy set should be assigned to.

Is this better?

> | `data.relationships.exclusions`               | array\[object] | (previous value) | An array of references to excluded workspaces that the policy set should be assigned to. Sending an empty array clears all exclusions assignments. You can only specify this attribute when `data.attributes.global` is `false`.                                                                                      |                                                                                                                                          |

Reviewer: ```suggestion
| `data.relationships.exclusions`               | array\[object] | (previous value) | An array of references to workspaces that the policy set should not be assigned to. Sending an empty array clears all exclusions assignments. You can only specify this attribute when `data.attributes.global` is `false`.                                                                                      |                                                                                                                                          |
```

  https://app.terraform.io/api/v2/policy-sets/polset-3yVQZvHzf5j3WRJ1/relationships/policies
```

> ## Attach a Policy Set to Projects

Author: @rkoron007 fixed Capitalization of the titles to make them consistent.

Reviewer 2: Thank you so much for putting in the effort here.  We've recently [updated the style guide](https://github.com/hashicorp/engineering-docs/blob/main/writing/style-guide.md#titles-and-headers) to specify that headers should use sentence casing. It's a not a huge priority, but it you are willing to make the change across the page to ensure all headings use sentence casing instead that would be **amazing**! No problem if you'd rather not though, just let me know and I can handle it later!

We also don't capitalize feature names, just _product_ names (so we capitalize Terraform Cloud, but not projects which are a _feature_ of Terraform Cloud). So we'd ideally update this to:
```suggestion
## Attach a policy set to projects
```

Author: I have fixed it in all places on this page.

  https://app.terraform.io/api/v2/policy-sets/polset-3yVQZvHzf5j3WRJ1/relationships/workspaces
```

> ## Attach a Policy Set to exclusions

Reviewer 2: ```suggestion
## Exclude a workspace from a policy set 
```

Reviewer 2: On further thought, I think we should use "exclude"/ "reinclude" here to explain how a workspace is "Excluded from a policy set's enforcement", which can be undone by "reincluding a workspace to a policy set's enforcement".  What do you think?

Reviewer 2: Or does this both "attach" and "exclude"? Sorry to nitpick, negative phrasing is so confusing by default 😅

Author: Exclude sounds good!

> `POST /policy-sets/:id/relationships/excludedWorkspaces`

Reviewer: Is the endpoint called `excludedWorkspaces` or `excluded`?

Author: it's exclusions now, fixed!

| Parameter | Description                                                                                                 |
| --------- | ----------------------------------------------------------------------------------------------------------- |

> | `:id`     | The ID of the policy set to attach to excluded workspaces. Use the "List Policy Sets" endpoint to find IDs. |

Reviewer 2: The phrasing here is a little hard to follow. Is the idea that you pass Terraform Cloud the id of a policy set _and_ the workspaces it should exclude from that set? I tried rephrasing, but I'm not sure if this is accurate?
```suggestion
| `:id`     | The ID of a policy set that you want Terraform Cloud to exclude from the workspaces you specify. Use the "List Policy Sets" endpoint to find IDs. |
```

Reviewer 2: (gentle bump on this, do you think the above suggestion works?)

Author: This is accurate, but the wording can be perhaps simplified?
How about this: The ID of a policy set that you want to exclude the specified workspaces from. What do you think about this?

Author: @rkoron007 in case you missed it, any thoughts on this?

Author: On second thought, I think this works.

Author: fixed!

  https://app.terraform.io/api/v2/policy-sets/polset-3yVQZvHzf5j3WRJ1/relationships/workspaces
```

> ## Detach the Policy Set from exclusions

Reviewer 2: ```suggestion
## Reinclude a workspace to a policy set
```

`DELETE /policy-sets/:id/relationships/excludedWorkspaces`

| Parameter | Description                                                                                                   |
| --------- | ------------------------------------------------------------------------------------------------------------- |

> | `:id`     | The ID of the policy set to detach from excluded workspaces. Use the "List Policy Sets" endpoint to find IDs. |

Reviewer 2: The negative phrasing is a bit confusing here. This effectively just returns a policy set's enforcement back to normal state, right? Can you please add something to that effect?

Author: @rkoron007 would this be better: The ID of the policy set reattach to excluded workspaces.

Reviewer 2: I think the verb "reattach" feels confusing here. It's excluded and reincluded, right?

```suggestion
| `:id`     | The ID of the policy set Terraform Cloud should reinclude (enforce) on the specified workspaces. Use the "List Policy Sets" endpoint to find IDs. |
```

### Request Body

This DELETE endpoint requires a JSON object with the following properties as a request payload.

Properties without a default value are required.

| Key path | Type           | Default | Description                                                                                                                                                                                                                                                                                                                                                                                                                                         |
| -------- | -------------- | ------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |

> | `data[]` | array\[object] |         | A list of resource identifier objects that defines which excluded workspaces the policy set will be detached from. These objects must contain `id` and `type` properties, and the `type` property must be `workspaces` (e.g. `{ "id": "ws-2HRvNs49EWPjDqT1", "type": "workspaces" }`). Obtain workspace IDs from the [workspace settings](/terraform/cloud-docs/workspaces/settings) or the [Show Workspace](/terraform/cloud-docs/api-docs/workspaces#show-workspace) endpoint. |

Reviewer 2: ```suggestion
| `data[]` | array\[object] |         | A list of resource identifier objects that defines which workspaces Terraform Cloud should reinclude (enforce) this policy set on. These objects must contain `id` and `type` properties, and the `type` property must be `workspaces` (e.g. `{ "id": "ws-2HRvNs49EWPjDqT1", "type": "workspaces" }`). Obtain workspace IDs from the [workspace settings](/terraform/cloud-docs/workspaces/settings) or the [Show Workspace](/terraform/cloud-docs/api-docs/workspaces#show-workspace) endpoint. |
```

| `newest-version`  | The most recently created policy set version, regardless of status. Note that this relationship may include an errored and unusable version, and is intended to allow checking for VCS errors. |
| `organization`    | The organization associated with the specified policy set.                                                                                                                                     |
| `policies`        | Individually managed policies which are associated with the policy set.                                                                                                                        |
| `projects`        | The projects this policy set applies to.                                                                                                                                                  |
| `workspaces`      | The workspaces this policy set applies to. 

> | `exclusions`      | The workspaces excluded from to this policy set's enforcement.                                                                                                                                                 |

Reviewer 2: ```suggestion
| `exclusions`      | The workspaces excluded from a policy set's enforcement.                                                                                                                                                 |
```

| `organization`    | The organization associated with the specified policy set.                                                                                                                                     |
| `policies`        | Individually managed policies which are associated with the policy set.                                                                                                                        |

> | `projects`        | The project this policy set applies to.                                                                                                                                                  |

Reviewer: Something we missed previously:

```suggestion
| `projects`        | The projects this policy set applies to.                                                                                                                                                  |
```

> | `workspaces`      | The workspaces to which the policy set applies. 
> | `exclusions`      | The excluded workspaces to which the policy set applies.                                                                                                                                                |

Reviewer: ```suggestion
| `workspaces`      | The workspaces this policy set applies to. 
| `exclusions`      | The workspaces excluded from to this policy set's enforcement.                                                                                                                                                |
```
